cmd/kubeadm/app/cmd/phases/uploadconfig.go

```diff
@@ -40,49 +40,109 @@ var (
+	Cfg() *kubeadmapi.InitConfiguration
+	Client() (clientset.Interface, error)
+}
+
+// NewUploadConfigPhase returns the phase to uploadConfig
+func NewUploadConfigPhase() workflow.Phase {
+	return workflow.Phase{
+		Name:    "upload-config",
 		Aliases: []string{"uploadconfig"},
-		Run: func(_ *cobra.Command, args []string) {
-			if len(cfgPath) == 0 {
-				kubeadmutil.CheckErr(errors.New("the --config flag is mandatory"))
-			}
+		Short:   "Uploads kubeadm and kubelet configuration to a ConfigMap",
+		Long:    cmdutil.MacroCommandLongDescription,
+		Phases: []workflow.Phase{
+			{
+				Name:     "kubeadm",
+				Short:    "Uploads the kubeadm ClusterConfiguration to a ConfigMap",
+				Long:     uploadConfigLongDesc,
+				Example:  uploadConfigExample,
+				Run:      runUploadKubeadmConfig,
+				CmdFlags: getUploadPhaseFlags(),
+			},
+			{
+				Name:     "kubelet",
+				Short:    "Uploads kubelet component config to a ConfigMap",
+				Long:     kubeletConfigUploadLongDesc,
+				Example:  kubeletConfigUploadExample,
+				Run:      runUploadKubeletConfig,
+				CmdFlags: getUploadPhaseFlags(),
+			},
+		},
+	}
+}
+
+func getUploadPhaseFlags() []string {
+	return []string{
+		options.CfgPath,
+		options.KubeconfigPath,
+	}
+}
 
-			kubeConfigFile = cmdutil.FindExistingKubeConfig(kubeConfigFile)
-			client, err := kubeconfigutil.ClientSetFromFile(kubeConfigFile)
-			kubeadmutil.CheckErr(err)
+// runUploadKubeadmConfig uploads the kubeadm configuration to a ConfigMap
+func runUploadKubeadmConfig(c workflow.RunData) error {
+	cfg, client, err := getUploadConfigData(c)
+	if err != nil {
+		return err
+	}
 
-			// KubernetesVersion is not used, but we set it explicitly to avoid the lookup
-			// of the version from the internet when executing ConfigFileAndDefaultsToInternalConfig
-			SetKubernetesVersion(cfg)
+	glog.V(1).Infof("[init] uploading currently used configuration to the cluster")
+	if err := uploadconfig.UploadConfiguration(cfg, client); err != nil {
+		return errors.Wrap(err, "error uploading configuration")
+	}
+	return nil
+}
 
-			internalcfg, err := configutil.ConfigFileAndDefaultsToInternalConfig(cfgPath, cfg)
-			kubeadmutil.CheckErr(err)
+// runUploadKubeletConfig uploads the kubelet configuration to a ConfigMap
+func runUploadKubeletConfig(c workflow.RunData) error {
+	cfg, client, err := getUploadConfigData(c)
+	if err != nil {
+		return err
+	}
 
-			err = uploadconfig.UploadConfiguration(internalcfg, client)
-			kubeadmutil.CheckErr(err)
-		},
+	glog.V(1).Infof("[init] creating kubelet configuration configmap")
+	if err = kubeletphase.CreateConfigMap(cfg, client); err != nil {
+		return errors.Wrap(err, "error creating kubelet configuration ConfigMap")
 	}
 
-	options.AddKubeConfigFlag(cmd.Flags(), &kubeConfigFile)
-	cmd.Flags().StringVar(&cfgPath, "config", "", "Path to a kubeadm config file. WARNING: Usage of a configuration file is experimental")
+	glog.V(1).Infof("[init] preserving the crisocket information for the master")
```

Comment:
`[upload-config] Preserving the CRISocket information for the control-plane node`